### 텐서플로 2.0 시작하기 : 초보자용

In [58]:
!pip install -q tensorflow-gpu==2.0.0-rc1
import tensorflow as tf

# 데이터 가져오기

MNIST(손글씨 숫자(0,1,2 등) 데이터) 데이터셋을 로드하여 준비. 


> MNIST 데이터셋을 import해서 x_train의 shape를 살펴보면 (60000, 28, 28)인데 
이것의 의미는 60,000개의 데이터로 구성되어 있고, (이미지 데이터이므로) 해상도(28x28 픽셀)를 표현하는 NumPy 배열의 데이터라는 것. x_test(10000, 28, 28)는 10,000개의 데이터로 구성되어 있음.


> y_train의 shape는 (60000,)인데 60,000개의 데이터로, 손글씨 숫자 데이터이기 때문에 label 값은 0~9까지의 정수 배열을 가짐. 마찬가지로, y_test는 10,000개의 데이터로 구성되어 있음.



텐서플로 모델을 만들고 훈련할 수 있는 고수준 API인 tf.keras를 이용.

In [59]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()

# 데이터 탐색

In [60]:
x_train

array([[[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       ...,

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 

In [61]:
x_train.shape

(60000, 28, 28)

In [62]:
y_train

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [63]:
import numpy as np
np.unique(y_train)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=uint8)

In [64]:
y_train.shape

(60000,)

In [65]:
x_test.shape

(10000, 28, 28)

In [66]:
y_test.shape

(10000,)

# 데이터 전처리

샘플 값을 정수에서 부동소수로 변환.

255.0으로 나누는 이유는, train 세트의 이미지를 보면 픽셀 값의 범위가 0 - 255 사이이므로 신경망 모델에 주입하기 전에 이 값의 범위를 0 - 1 사이로 조정해야 하기 때문. train 세트와 test 세트를 동일한 방식으로 전처리하는 것이 중요!

In [67]:
x_train, x_test = x_train / 255.0, x_test / 255.0

In [68]:
x_train

array([[[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]],

       ...,

       [[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0.

# 모델 구성

층을 차례대로 쌓아 tf.keras.Sequential 모델을 만듦. 훈련에 사용할 optimizer와 loss function을 선택.

> Keras에서는 층(layer)을 조합하여 model을 만드는데, 모델은 일반적으로 층(layer)의 그래프. 가장 흔한 모델 구조는 층을 차례대로 쌓은 Sequential 모델.

> tf.keras.layers.Dense와 같은 층들의 가중치(parameter)는 훈련하는 동안 학습됨.





tf.keras.layers.Dense 층을 밀집 연결(densely-connected) 또는 완전 연결(fully-connected) 층이라고 부름. 첫 번째 Dense층은 128개의 노드를 가지고, 마지막 층은 10개의 노드의 softmax 층임. softmax 층은 10개의 확률을 반환하고, 반환된 값의 전체 합은 1임. 각 노드는 현재 이미지가 10개 클래스 중 하나에 속할 확률을 출력.

In [69]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28,28)),  # Flatten 메서드를 사용하여 2차원 배열(28x28 픽셀)의 이미지 포맷을 28*28=784 픽셀의 1차원 배열로 일렬로 변환.
    tf.keras.layers.Dense(128, activation='relu'), # 128개의 유닛을 가진 완전 연결 층을 모델에 추가(activation function은 Relu를 이용)
    tf.keras.layers.Dropout(0.2),  # 학습하는 매번 iteration마다 랜덤하게 20%를 dropout하는 regularization 추가
    tf.keras.layers.Dense(10, activation='softmax')  # 10개의 출력 유닛을 가진 소프트맥스 층을 추가
])

# 모델 컴파일

모델을 훈련하기 전 필요한 몇 가지 설정이 model compile단계에서 추가됨.


1.   손실 함수(Loss function) : 훈련하는 동안 모델의 오차를 측정. 모델의 학습이 올바른 방향으로 향하도록 이 함수를 최소화해야 함.
2.   Optimizer : 데이터와 손실 함수를 바탕으로 모델의 업데이트 방법을 결정.
3.   지표(metrics) : 훈련 단계와 테스트 단계를 모니터링하기 위해 사용함. 여기서는 올바르게 분류된 이미지의 비율인 '정확도'를 사용함.



In [70]:
model.compile(optimizer = 'adam',  # 기본적인 adam 모델
              loss = 'sparse_categorical_crossentropy',  # 추정 정보량의 기대값을 사용하는 cross entropy loss function
              metrics = ['accuracy'])  

# 모델 훈련

신경망 모델을 훈련하는 단계


1.   훈련 데이터(x_train, y_train)를 모델에 주입
2.   모델이 이미지와 label을 매핑하는 방법을 학습
3.   테스트 데이터(x_test)에 대해 모델이 예측하고, 이 예측이 (y_test)의 label과 맞는지 확인



tf.keras.Model.fit의 기본적인 매개변수


1.   epochs : 훈련은 epoch로 구성되는데, 한 epoch는 전체 입력 데이터를 한번 순회하는 것. (작은 배치로 나누어 수행)
2.   batch_size : 모델은 데이터를 작은 배치로 나누고 훈련 과정에서 이 배치를 순회. 정수 값으로 배치의 크기를 지정할 수 있는데, 전체 샘플 수가 배치 크기로 나누어 떨어지지 않으면 마지막 배치의 크기는 더 작을 수 있음.



In [71]:
model.fit(x_train, y_train, epochs = 5)

Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 9s 157us/sample - loss: 0.3007 - accuracy: 0.9132
Epoch 2/5
60000/60000 [==============================] - 7s 120us/sample - loss: 0.1454 - accuracy: 0.9564
Epoch 3/5
60000/60000 [==============================] - 7s 112us/sample - loss: 0.1100 - accuracy: 0.9670
Epoch 4/5
60000/60000 [==============================] - 7s 118us/sample - loss: 0.0890 - accuracy: 0.9725
Epoch 5/5
60000/60000 [==============================] - 8s 132us/sample - loss: 0.0755 - accuracy: 0.9766


# 모델 평가

In [72]:
# verbose : 얼마나 자세하게 정보를 표시할 것인가를 지정 (0, 1, 2)
# verbose = 2 : 진행막대 표시하지 않음
model.evaluate(x_test, y_test, verbose = 2)

10000/1 - 1s - loss: 0.0396 - accuracy: 0.9780


[0.0697683941679541, 0.978]

In [73]:
# verbose = 0 : 아무것도 출력하지 않음
model.evaluate(x_test, y_test, verbose = 0)

[0.0697683941679541, 0.978]

In [74]:
# verbose = 1 : 훈련의 진행도를 보여주는 진행 막대를 보여줌
model.evaluate(x_test, y_test, verbose = 1)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

[0.0697683941679541, 0.978]

모델이 훈련되면서 손실과 정확도 지표가 출력. 여기서 훈련된 이미지 분류기는 이 데이터셋에서 약 98%의 정확도를 달성.